File: 2020/day_3/day_3.py
"""Puzzle Day 3"""
import math
from pathlib import Path
from typing import List, Tuple


def read_in_map_slice():
    """The map is represented by '#' which are trees and '.' which are open spaces."""
    with open(Path(__file__).parent.joinpath("input.txt")) as puzzle_input:
        return [
            [character for character in line]
            for line in puzzle_input.read().splitlines()
        ]


def update_position(position, map_width, map_length, slope):
    """Update the current position based off the slope, wrap when at the edge of the map."""
    if position["x"] + slope[0] > map_width:
        position["x"] = (position["x"] + slope[0]) - map_width - 1
    else:
        position["x"] = position["x"] + slope[0]
    if position["y"] + slope[1] > map_length:
        position["y"] = map_length - 1
    else:
        position["y"] = position["y"] + slope[1]
    return position


def navigate_terrain(slope: Tuple[int, int], terrain_map: List[List[str]]) -> int:
    """Until we reach the bottom of the map, count the trees as we go."""
    num_of_trees = 0
    MAP_LENGTH = len(terrain_map) - 1
    MAP_WIDTH = len(terrain_map[0]) - 1
    position = {"x": 0, "y": 0}  # The upper left most corner.
    while position["y"] < MAP_LENGTH:
        position = update_position(position, MAP_WIDTH, MAP_LENGTH, slope)
        if terrain_map[position["y"]][position["x"]] == "#":
            num_of_trees += 1
    return num_of_trees


if __name__ == "__main__":
    terrain_map = read_in_map_slice()

    print(f"Part 1 Number of Trees: {navigate_terrain((3,1), terrain_map)}")
    # Part 1 Number of Trees: 289
    slopes = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]
    trees = [navigate_terrain(slope, terrain_map) for slope in slopes]
    print(f"Part 2 Number of Trees: {math.prod(trees)}")
    # Part 2 Number of Trees: 5522401584

File: 2020/day_3/day_3.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>

std::vector<std::string> read_in_map_slice( const std::string filename );
int navigate_terrain_and_count_trees( const std::vector<std::string> &terrain, const std::vector<int> &slope );
void update_position( std::vector<int> &position, int width, int length, const std::vector<int> &slope );


// Reading map file: input.txt
// Part 1: Tree Count: 289
// Part 2: Tree Count: 5522401584
int main()
{
    std::vector<std::string> terrain_map {read_in_map_slice( "input.txt" )};

    std::cout << "Part 1: Tree Count: " << navigate_terrain_and_count_trees( terrain_map, {3, 1} ) << std::endl;

    std::vector<std::vector<int>> slopes
    {
        {1, 1},
        {3, 1},
        {5, 1},
        {7, 1},
        {1, 2}
    };
    std::vector<int> tree_counts {};
    for( auto slope : slopes )
    {
        tree_counts.push_back( navigate_terrain_and_count_trees( terrain_map, slope ) );
    }
    size_t product {1};
    for( auto item : tree_counts )
    {
        product *= item;
    }
    std::cout << "Part 2: Tree Count: " << product << std::endl;
    return 0;
}

/**
 * Read in every line of the file into a vector.  The map is represented with '#' as trees and
 * '.' as open space.
 * */
std::vector<std::string> read_in_map_slice( const std::string filename )
{
    std::vector<std::string> terrain;
    std::cout << "Reading map file: " << filename << std::endl;
    std::ifstream file( filename );
    std::string map_segment{};
    while( std::getline( file, map_segment ) )
    {
        terrain.push_back( map_segment );
    }
    return terrain;
}

int navigate_terrain_and_count_trees( const std::vector<std::string> &terrain, const std::vector<int> &slope )
{
    int num_of_trees {0};
    std::vector<int> position {0, 0};
    constexpr int X {0};
    constexpr int Y {1};
    while( position.at( Y ) < terrain.size() - 1 )
    {
        update_position( position, terrain.at( 0 ).size() - 1, terrain.size() - 1, slope );
        if( terrain.at( position.at( Y ) ).at( position.at( X ) ) == '#' )
        {
            num_of_trees++;
        }
    }
    return num_of_trees;
}

/**
 *  Update the current position based off the slope, wrap when at the edge of the map.
 * */
void update_position( std::vector<int> &position, int width, int length, const std::vector<int> &slope )
{
    constexpr int X {0};
    constexpr int Y {1};
    if( position.at( X ) + slope.at( X ) > width )
    {
        position.at( X ) = ( position.at( X ) + slope.at( X ) ) - width - 1;
    }
    else
    {
        position.at( X ) = position.at( X ) + slope.at( X );
    }
    if( position.at( Y ) + slope.at( Y ) > length )
    {
        position.at( Y ) = length - 1;
    }
    else
    {
        position.at( Y ) = position.at( Y ) + slope.at( Y );
    }
}
